native/cocos/core/Root.cpp

```diff
@@ -69,7 +73,40 @@ Root::~Root() {
 }
 
 void Root::initialize(gfx::Swapchain *swapchain) {
-    _swapchain = swapchain;
+    //_swapchain = swapchain;
+
+    ISystemWindowManager *windowMgr = CC_GET_PLATFORM_INTERFACE(ISystemWindowManager);
+    const auto &windows = windowMgr->getWindows();
+    for (auto pair : windows) {
+        auto *window = pair.second.get();
+        _mainWindow = createRenderWindowFromSystemWindow(window);
+    }
+    _curWindow = _mainWindow;
+
+    // TODO(minggo):
+    // return Promise.resolve(builtinResMgr.initBuiltinRes(this._device));
+}
+
+render::Pipeline *Root::getCustomPipeline() const {
+    return dynamic_cast<render::Pipeline *>(_pipelineRuntime.get());
+}
+
+scene::RenderWindow *Root::createRenderWindowFromSystemWindow(ISystemWindow *window) {
+    if (!window)
+        return nullptr;
+
+    uint32_t windowId = window->getWindowId();
+    auto handle = window->getWindowHandle();
+    const auto &size = window->getViewSize();
+
+    gfx::SwapchainInfo info;
+    info.width  = size.x;
+    info.height = size.y;
+    info.windowHandle = reinterpret_cast<void *>(handle);
+    info.windowId = window->getWindowId();
+
+    gfx::Swapchain *swapchain = gfx::Device::getInstance()->createSwapchain(info);
+    _swapchains.emplace_back(swapchain);
```

Comment:
Why need record swapchains in Root?

Reply:
`_swapchains ` is used in Root::frameMove
`_device->acquire(_swapchains);`

Reply:
Ok.